src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -1557,6 +1557,510 @@ public static void Main()
+            comp.VerifyDiagnostics(
+                // (8,26): warning CS8601: Possible null reference assignment.
+                //         _ = b with { X = s }; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "s").WithLocation(8, 26));
+        }
+
+        [Fact]
+        public void WithExpr_NullableAnalysis_03()
+        {
+            var src = @"
+#nullable enable
+data class B(string? X)
+{
+    public B Clone() => new B(X);
+
+    static void M1(B b, string s, bool flag)
+    {
+        if (flag) { b.X.ToString(); } // 1
+        _ = b with { X = s };
+        if (flag) { b.X.ToString(); } // 2
+    }
+
+    static void M2(B b, string s, bool flag)
+    {
+        if (flag) { b.X.ToString(); } // 3
+        b = b with { X = s };
+        if (flag) { b.X.ToString(); }
+    }
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (9,21): warning CS8602: Dereference of a possibly null reference.
+                //         if (flag) { b.X.ToString(); } // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.X").WithLocation(9, 21),
+                // (11,21): warning CS8602: Dereference of a possibly null reference.
+                //         if (flag) { b.X.ToString(); } // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.X").WithLocation(11, 21),
+                // (16,21): warning CS8602: Dereference of a possibly null reference.
+                //         if (flag) { b.X.ToString(); } // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.X").WithLocation(16, 21));
+        }
+
+        [Fact]
+        public void WithExpr_NullableAnalysis_04()
+        {
+            var src = @"
+#nullable enable
+data class B(int X)
+{
+    static void M1(B? b)
+    {
+        var b1 = b with { X = 42 }; // 1
+        _ = b.ToString();
+        _ = b1.ToString();
+    }
+
+    static void M2(B? b)
+    {
+        (b with { X = 42 }).ToString(); // 2
+    }
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (7,18): warning CS8602: Dereference of a possibly null reference.
+                //         var b1 = b with { X = 42 }; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b").WithLocation(7, 18),
+                // (14,10): warning CS8602: Dereference of a possibly null reference.
+                //         (b with { X = 42 }).ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b").WithLocation(14, 10));
+        }
+
+        [Fact]
+        public void WithExpr_NullableAnalysis_05()
+        {
+            var src = @"
+#nullable enable
+data class B(string? X, string? Y)
+{
+    public B Clone() => new B(X, Y);
+
+    static void M1(bool flag)
+    {
+        B b = new B(""hello"", null);
+        if (flag)
+        {
+            b.X.ToString(); // PROTOTYPE(records)
```

Comment:
>// PROTOTYPE(records) [](start = 28, length = 21)

Please replace PROTOTYPE with link to issue.

Reply:
Done